tests/qos/test_pfc_counters.py

```diff
@@ -0,0 +1,144 @@
+        peer_port = conn_facts[intf]['peerport']
+        peer_port_name = eos_to_linux_intf(peer_port)
+
+        peerdev_ans = AnsibleHost(ansible_adhoc, peer_device)
+        for priority in range(PRIO_COUNT):
+            cmd = "sudo python %s -i %s -p %d -t %d -n %d" % (PFC_GEN_FILE_DEST, peer_port_name, 2 ** priority, pause_time, PKT_COUNT)
+            peerdev_ans.command(cmd)
+	
+    """ SONiC takes some time to update counters in database """
+    time.sleep(5)
+
+    """ Check results """
+    dut_hostname = testbed['dut']
+    dut_ans = AnsibleHost(ansible_adhoc, dut_hostname)
+    counter_facts = dut_ans.sonic_pfc_counters(method = "get")['ansible_facts']
+
+    for intf in conn_facts:
+        assert intf in counter_facts
+        assert 'Rx' in counter_facts[intf]
+        assert counter_facts[intf]['Rx'] == [str(PKT_COUNT)] * PRIO_COUNT
+
+def run_fc(ansible_adhoc, testbed, conn_graph_facts, pause_time):
+    """
+    @Summary: Run the flow control (FC) test case
+    @param ansible_adhoc: Fixture provided by the pytest-ansible package. Source of the various device objects. It is
+    mandatory argument for the class constructors.
+    @param testbed: Testbed information
+    @param conn_graph_facts: Testbed topology connectivity information
+    @param pause_time: Pause time quanta (0-65535) in the frame. 0 means unpause.
+    """
+    setup_testbed(ansible_adhoc, testbed, conn_graph_facts)
+    conn_facts = conn_graph_facts['device_conn']
+
+    """ Generate flow control packets for all the interfaces """
+    for intf in conn_facts:
+        peer_device = conn_facts[intf]['peerdevice']
+        peer_port = conn_facts[intf]['peerport']
+        peer_port_name = eos_to_linux_intf(peer_port)
+
+        peerdev_ans = AnsibleHost(ansible_adhoc, peer_device)
+        cmd = "sudo python %s -i %s -g -t %d -n %d" % (PFC_GEN_FILE_DEST, peer_port_name, pause_time, PKT_COUNT)
+        peerdev_ans.command(cmd)
+	
+    """ SONiC takes some time to update counters in database """
+    time.sleep(5)
+
+    """ Check resuls. SONiC should not update PFC counters when receiving global pause/unpause frames """
+    dut_hostname = testbed['dut']
+    dut_ans = AnsibleHost(ansible_adhoc, dut_hostname)
+    counter_facts = dut_ans.sonic_pfc_counters(method = "get")['ansible_facts']
+    
+    for intf in conn_facts:
+        assert intf in counter_facts
+        assert 'Rx' in counter_facts[intf]
+        assert counter_facts[intf]['Rx'] == ['0'] * PRIO_COUNT
+
+def eos_to_linux_intf(eos_intf_name):
+    """
+    @Summary: Map EOS's interface name to Linux's interface name
+    @param eos_intf_name: Interface name in EOS
+    @return: Return the interface name in Linux 
+    """
+    return eos_intf_name.replace('Ethernet', 'et').replace('/', '_')
+
+def test_pfc_pause(ansible_adhoc, testbed, conn_graph_facts):
+    """ @Summary: Run PFC pause frame (pause time quanta > 0) tests """
+    run_pfc(ansible_adhoc, testbed, conn_graph_facts, 65535)
```

Comment:
Can you try merging the run_pfc and run_fc? There is a lot of code duplication in both. The only difference is enabling the '-g' to send global pause and counter_facts check. You can do something like this by passing a bool for global pause.

if global pause:
     global_pause = '-g'
     counter_values = ['0'] * PRIO_COUNT
else:
     global_pause = ''
     counter_ values = [str(PKT_COUNT)] * PRIO_COUNT

Reply:
@neethajohn I think run_pfc and run_fc are a little bit different. In addition to '-g' option, run_pfc needs to generate PFC frames on all the priorities while run_fc does not.

Reply:
@neethajohn  Fixed! Merge two functions into run_test